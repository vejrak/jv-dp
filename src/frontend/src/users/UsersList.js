// @flow

import React, { Component } from 'react'
import { Button } from 'react-bootstrap'
import type { ErrorResponse, User } from '../types'
import { getUsers, deleteUser } from './actions'
import Spinner from '../components/Spinner'

type Props = $ReadOnly<{|
  isFetching: boolean,
  users: Array<User>,
  getUsers: typeof getUsers,
  error: ?ErrorResponse,
  deleteUser: typeof deleteUser,
|}>

class UsersList extends Component<Props> {
  componentDidMount() {
    this.props.getUsers()
  }

  handleUserDelete = async (_id: string) => {
    await this.props.deleteUser(_id)
  }

  render() {
    const { error, users, isFetching } = this.props
    if (isFetching) return <Spinner />

    if (error)
      return (
        <div className="alert alert-danger" role="alert">
          {error.errmsg}
        </div>
      )

    return (
      <div>
        {users.length > 0 && (
          <table className="table">
            <tbody>
              <tr>
                <th>Username</th>
                <th>Rank</th>
                <th />
              </tr>
            </tbody>
            {users.map((value) => (
              <tbody key={value._id}>
                <tr>
                  <td className="col">{value.username}</td>
                  <td className="col">{value.rank}</td>
                  <td className="col">
                    <Button
                      className="btn d-block btn-danger"
                      disabled={isFetching}
                      onClick={() => {
                        if (
                          window.confirm(
                            'Are you sure you wish to delete this item?',
                          )
                        )
                          this.handleUserDelete(value._id)
                      }}
                    >
                      Delete
                    </Button>
                  </td>
                </tr>
              </tbody>
            ))}
          </table>
        )}
      </div>
    )
  }
}

export default UsersList
