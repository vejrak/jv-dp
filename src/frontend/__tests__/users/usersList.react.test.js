import React from 'react'
import renderer from 'react-test-renderer'
import UsersList from '../../src/users/UsersList'

const mockProps = {
  getUsers: () => {},
  deleteUser: () => {},
  isFetching: false,
  error: null,
  users: [{ _id: 'id1', username: 'name1', rank: 1 }],
}

it('renders correctly', () => {
  const tree = renderer.create(<UsersList {...mockProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
