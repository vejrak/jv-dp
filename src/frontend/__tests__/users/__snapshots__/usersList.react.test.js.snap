// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <th>
          Username
        </th>
        <th>
          Rank
        </th>
        <th />
      </tr>
    </tbody>
    <tbody>
      <tr>
        <td
          className="col"
        >
          name1
        </td>
        <td
          className="col"
        >
          1
        </td>
        <td
          className="col"
        >
          <button
            className="btn d-block btn-danger btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
